File: akka-sample-cluster-client-grpc-scala/src/main/scala/sample/cluster/client/grpc/ClusterClientReceptionistGrpcImpl.scala
package sample.cluster.client.grpc

import java.util.UUID

import akka.NotUsed
import akka.actor.ActorRef
import akka.cluster.pubsub.DistributedPubSubMediator
import akka.event.LoggingAdapter
import akka.stream.{Materializer, OverflowStrategy}
import akka.stream.scaladsl.Source
import akka.util.Timeout

import scala.concurrent.Future
import scala.util.control.NonFatal

class ClusterClientReceptionistGrpcImpl(
  settings: ClusterReceptionistSettings,
  pubSubMediator: ActorRef,
  serialization: ClusterClientSerialization
)(implicit
  mat: Materializer,
  log: LoggingAdapter)
    extends ClusterClientReceptionistService {

  override def newSession(in: Source[Req, NotUsed]): Source[Rsp, NotUsed] = {
    val sessionId = UUID.randomUUID().toString
    log.info("New session [{}]", sessionId)
    Source
      .actorRef[Any](
        bufferSize = settings.bufferSize,
        overflowStrategy = OverflowStrategy.dropNew,
        // never complete from stream element
        completionMatcher = PartialFunction.empty,
        // never fail from stream element
        failureMatcher = PartialFunction.empty
      )
      .map { rsp =>
        val payload = serialization.serializePayload(rsp)
        Rsp(Some(payload))
      }
      .mapMaterializedValue { sessionRspRef =>
        in.runForeach { req =>
            if (req.req.isSend) {
              val sendReq = req.getSend
              val msg = serialization.deserializePayload(sendReq.payload.get)
              // using sessionRspRef as sender so that replies are emitted to the response stream back to the client
              pubSubMediator.tell(
                DistributedPubSubMediator
                  .Send(sendReq.path, msg, sendReq.localAffinity),
                sessionRspRef
              )
            } else if (req.req.isSendToAll) {
              val sendToAllReq = req.getSendToAll
              val msg =
                serialization.deserializePayload(sendToAllReq.payload.get)
              pubSubMediator.tell(
                DistributedPubSubMediator.SendToAll(sendToAllReq.path, msg),
                sessionRspRef
              )
            } else if (req.req.isPublish) {
              val publishReq = req.getPublish
              val msg = serialization.deserializePayload(publishReq.payload.get)
              pubSubMediator.tell(
                DistributedPubSubMediator.Publish(publishReq.topic, msg),
                sessionRspRef
              )
            } else {
              throw new IllegalArgumentException("Unknown request type")
            }

          }
          .onComplete { result =>
            log.info("Session [{}] completed: {}", sessionId, result)
          }(mat.executionContext)
        NotUsed
      }
  }

  override def askSend(sendReq: SendReq): Future[Rsp] = {
    try {
      import akka.pattern.ask
      implicit val timeout = Timeout(settings.askSendTimeout)
      implicit val ec = mat.executionContext
      val msg = serialization.deserializePayload(sendReq.payload.get)
      (pubSubMediator ? DistributedPubSubMediator.Send(
        sendReq.path,
        msg,
        sendReq.localAffinity
      )).map { rsp =>
        val payload = serialization.serializePayload(rsp)
        Rsp(Some(payload))
      }
    } catch {
      case NonFatal(e) =>
        // deserialization error
        Future.failed(e)

    }
  }
}

File: akka-sample-cluster-client-grpc-java/src/main/java/sample/cluster/client/grpc/ClusterClientReceptionistGrpcImpl.java
package sample.cluster.client.grpc;

import akka.NotUsed;
import akka.actor.ActorRef;
import akka.cluster.pubsub.DistributedPubSubMediator;
import akka.event.LoggingAdapter;
import akka.pattern.Patterns;
import akka.stream.Materializer;
import akka.stream.OverflowStrategy;
import akka.stream.javadsl.Source;

import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

class ClusterClientReceptionistGrpcImpl implements ClusterClientReceptionistService {
  private final ClusterReceptionistSettings settings;
  private final ActorRef pubSubMediator;
  private final ClusterClientSerialization serialization;
  private final Materializer materializer;
  private final LoggingAdapter log;

  ClusterClientReceptionistGrpcImpl(ClusterReceptionistSettings settings, ActorRef pubSubMediator, ClusterClientSerialization serialization, Materializer materializer, LoggingAdapter log) {
    this.settings = settings;
    this.pubSubMediator = pubSubMediator;
    this.serialization = serialization;
    this.materializer = materializer;
    this.log = log;
  }

  @Override
  public Source<Rsp, NotUsed> newSession(Source<Req, NotUsed> in) {
    final String sessionId = UUID.randomUUID().toString();
    log.info("New session [{}]", sessionId);
    return Source
      .actorRef(
        // never complete from stream element
        elem -> Optional.empty(),
        // never fail from stream element
        elem -> Optional.empty(),
        settings.bufferSize,
        OverflowStrategy.dropNew())
      .map( rsp -> {
        Payload payload = serialization.serializePayload(rsp);
        return Rsp.newBuilder().setPayload(payload).build();
      })
      .mapMaterializedValue( sessionRspRef -> {
        in.runForeach( req -> {
          if (req.getReqCase() == Req.ReqCase.SEND) {
            SendReq sendReq = req.getSend();
            Object msg = serialization.deserializePayload(sendReq.getPayload());
            // using sessionRspRef as sender so that replies are emitted to the response stream back to the client
            pubSubMediator.tell(
              new DistributedPubSubMediator.Send(sendReq.getPath(), msg, sendReq.getLocalAffinity()),
              sessionRspRef);
          } else if (req.getReqCase() == Req.ReqCase.SENDTOALL) {
            SendToAllReq sendToAllReq = req.getSendToAll();
            Object msg = serialization.deserializePayload(sendToAllReq.getPayload());
            pubSubMediator.tell(
              new DistributedPubSubMediator.SendToAll(sendToAllReq.getPath(), msg),
              sessionRspRef);
          } else if (req.getReqCase() == Req.ReqCase.PUBLISH) {
            PublishReq publishReq = req.getPublish();
            Object msg = serialization.deserializePayload(publishReq.getPayload());
            pubSubMediator.tell(
              new DistributedPubSubMediator.Publish(publishReq.getTopic(), msg),
              sessionRspRef);
          } else {
            throw new IllegalArgumentException("Unknown request type");
          }

        }, materializer)
          .whenComplete((result, exc) -> {
            if (exc == null)
              log.info("Session [{}] completed successfully: {}", sessionId, result);
            else
              log.info("Session [{}] completed with failure: {}", sessionId, exc);
          });

        return NotUsed.getInstance();
      });
  }

  @Override
  public CompletionStage<Rsp> askSend(SendReq sendReq) {
    try {
      Object msg = serialization.deserializePayload(sendReq.getPayload());
      return Patterns.ask(
        pubSubMediator,
        new DistributedPubSubMediator.Send(sendReq.getPath(), msg, sendReq.getLocalAffinity()),
        settings.askSendTimeout)
        .thenApply( rsp -> {
        Payload payload = serialization.serializePayload(rsp);
        return Rsp.newBuilder().setPayload(payload).build();
      });
    } catch (RuntimeException e) {
      // deserialization error
      CompletableFuture<Rsp> result = new CompletableFuture<>();
      result.completeExceptionally(e);
      return result;
    }
  }

}
